issue 1270 - added DAO factory pattern (#1920)
iluwatar/java-design-patterns: dao-factory/README.md

---
layout: pattern

> title: Dao Factory

Reviewer: ```suggestion
title: DAO Factory
```

folder: dao-factory
permalink: /patterns/dao-factory/
categories: Creational
language: en
tags:

> - Decoupling

Reviewer: I would add also `Data access`

---

## Intent

> The Dao Factory Pattern provides an abstract DAO factory object (Abstract Factory) that can construct various types of concrete DAO factories, each factory supporting a different type of persistent storage implementation. Once you obtain the concrete DAO factory for a specific implementation, you use it to produce DAOs supported and implemented in that implementation.

Reviewer: ```suggestion
The DAO Factory Pattern provides an abstract DAO factory object (Abstract Factory) that can construct various types of concrete DAO factories, each factory supporting a different type of persistent storage implementation. Once you obtain the concrete DAO factory for a specific implementation, you use it to produce DAOs supported and implemented in that implementation.
```

## Explanation

Real world example

> There are a few databases and if the implementation changes from Derby to another product, the only required change is the getDAOFactory() method call to the DAO factory to obtain a different factory.

> > 

Reviewer: Can you paint the picture on a little bit higher abstraction level e.g. there is an application that is used to route nuclear missiles to their targets. The application uses Mongo database in production, but the developers like to use Derby for easy development and testing. DAO Factory pattern enables switching the used database easily.

In plain words

> > By using this pattern, we can switch between underlying databases.

Reviewer: Use passive voice https://www.grammarly.com/blog/passive-voice/

> 
**Programmatic Example**

> Walking through our example, we have User entity `User`.

Reviewer: ```suggestion
Walking through our example, we have an entity called `User`.
```

```java
@Getter
@Setter
public class User implements java.io.Serializable {
    private int UserId;
    private String name;
    private String streetAddress;
    private String city;

}
```

> Here's the `UserDAO` interface that all UserDAOs must support. The DAO methods for User persistent object that are implemented by all concrete DAO implementations, such as DerbyUserDao and MongoUserDao.

Reviewer: ```suggestion
Here's the `UserDAO` interface that all DAOs must support. It contains methods for user persistent objects that are implemented by all concrete DAO implementations, such as DerbyUserDao and MongoUserDao.
```
